### PR TITLE
Advise `Vary` even for non-CORS-request responses

### PR DESCRIPTION
**Pull request by sideshowbarker**
See https://stackoverflow.com/a/45081016/441757

***
[Preview](https://fetch.spec.whatwg.org/branch-snapshots/sideshowbarker/vary-for-all-responses/) | [Diff](https://s3.amazonaws.com/pr-preview/whatwg/fetch/f3bb219...6d74cb5.html)

### CODE REVIEW

**Comment by tyoshino** [comment]
Thank you Michael. Looks good basically, but maybe for those who are not familiar with the issue, what "all" means is a little unclear? Can we just talk about the concrete story (a response to no-CORS request being cached and served for following CORS requests)?

Slight point, but I'd split the sentence, i.e. not "Vary is to be used for" but "Vary is to be used. Such a Vary header needs to be ...".

File: fetch.bs [hunk 1]
```diff
@@ -5658,11 +5658,12 @@ however, it is perfectly fine to do so.
...
 <p>If <a>CORS protocol</a> requirements are more complicated than setting
 `<a http-header><code>Access-Control-Allow-Origin</code></a>` to <code>*</code> or a static
-<a for=/>origin</a>, `<code>Vary</code>` is to be used.
+<a for=/>origin</a>, `<code>Vary</code>` is to be used for <em>all</em> responses for a particular
+resource — not just for responses to <a lt="CORS request">CORS requests</a> for that resource.
 [[!HTML]]
 [[!HTTP]] [[!HTTP-SEMANTICS]] [[!HTTP-COND]] [[!HTTP-CACHING]] [[!HTTP-AUTH]]
 
-<pre id=example-vary-origin class=example>Vary: Origin</pre>
+<pre id=example-vary-origin class=example>Vary: Origin, Access-Control-Request-Headers, Access-Control-Request-Method</pre>
```
Comment (annevk) [inline fetch.bs]: I don't think it makes sense to include those additional headers, as they're only relevant for the preflight. They're not included in the "actual" request.

Comment (jakearchibald) [inline fetch.bs]: Oh, yeah, good point. I guess you don't need `Access-Control-Allow-Credentials` in there as the credentials is part of the cache key.

Comment (annevk) [inline fetch.bs]: Yeah, though we haven't written that down yet and last I heard Chrome doesn't believe in that. :/

Comment (sideshowbarker) [inline fetch.bs]: > I don't think it makes sense to include those additional headers, as they're only relevant for the preflight. They're not included in the "actual" request.

OK, removed those

**Comment by sideshowbarker** [comment]
> maybe for those who are not familiar with the issue, what "all" means is a little unclear? Can we just talk about the concrete story (a response to no-CORS request being cached and served for following CORS requests)?

Yes—I’ve now pushed b98d51b to here, with additional wording that attempts to describe the concrete scenario much more explicitly.

**Comment by mnot** [comment]
Looks pretty good. Technically, `Vary` is only necessary when the response is cacheable, but that's often broader than most people realise.

**Comment by tyoshino** [comment]
b98d51b looks good

**Comment by annevk** [comment]
We should maybe also point out that if you intend to serve a static value such as a single origin or `*` you don't want to set the response header conditionally and don't want to use `Vary`.

**Comment by annevk** [comment]
(Especially if we add comma-separated origins at some point as a possible value.)

**Comment by sideshowbarker** [comment]
> We should maybe also point out that if you intend to serve a static value such as a single origin or \* you don't want to set the response header conditionally and don't want to use Vary.

OK, added in 5e33292

**Comment by tyoshino** [comment]
Still looks good. Nice to mention the case.

File: fetch.bs [hunk 2]
```diff
@@ -5664,7 +5664,26 @@ however, it is perfectly fine to do so.
...
+the user agent to <a for=/>fetch</a> a response that includes
+`<a http-header><code>Access-Control-Allow-Origin</code></a>`, rather than using the cached response
+from the previous non-<a>CORS request</a> that lacks
+`<a http-header><code>Access-Control-Allow-Origin</code></a>`.
+
+<p>However, if `<a http-header><code>Access-Control-Allow-Origin</code></a>` is set to
+<code>*</code> or a static <a for=/>origin</a> for a particular resource, then configure the server
+to always send `<a http-header><code>Access-Control-Allow-Origin</code></a>` in responses for the
+resource — for non-<a lt="CORS request">CORS requests</a> as well as <a lt="CORS request">CORS
+requests</a> — and do not use `<code>Vary</code>`.
```
Comment (annevk) [inline fetch.bs]: Extra newline here.

File: fetch.bs [hunk 3]
```diff
@@ -5664,7 +5664,26 @@ however, it is perfectly fine to do so.
...
+<p>But if `<code>Vary: Origin</code>` is used in the same scenario described above, it will cause
+the user agent to <a for=/>fetch</a> a response that includes
+`<a http-header><code>Access-Control-Allow-Origin</code></a>`, rather than using the cached response
+from the previous non-<a>CORS request</a> that lacks
+`<a http-header><code>Access-Control-Allow-Origin</code></a>`.
+
+<p>However, if `<a http-header><code>Access-Control-Allow-Origin</code></a>` is set to
+<code>*</code> or a static <a for=/>origin</a> for a particular resource, then configure the server
+to always send `<a http-header><code>Access-Control-Allow-Origin</code></a>` in responses for the
+resource — for non-<a lt="CORS request">CORS requests</a> as well as <a lt="CORS request">CORS
```
Comment (annevk) [inline fetch.bs]: No need for the `lt` attribute here. Bikeshed knows about English plural.

**Comment by annevk** [comment]
Thanks!

**Pull request merged**